src/tsickle.ts

```diff
@@ -853,13 +856,20 @@ class Annotator extends ClosureRewriter {
       }
       const resolved = ts.resolveModuleName(moduleId, this.file.fileName, this.tsOpts, this.host);
       if (resolved && resolved.resolvedModule) {
-        const resolvedModule = resolved.resolvedModule.resolvedFileName.replace(/(\.d)?\.ts$/, '');
-        const requestedModule = moduleId.replace(/\.js$/, '');
+        const requestedModule = moduleId.replace(EXT, '');
+        const resolvedModule = resolved.resolvedModule.resolvedFileName.replace(EXT, '');
         // If the imported module resolves to foo/index, but the specified module was foo, then we
         // append the /index.
-        if (resolvedModule.substr(resolvedModule.length - 6) === '/index' &&
-            requestedModule.substr(requestedModule.length - 6) !== '/index') {
-          moduleId += '/index';
+        if (requestedModule.substr(requestedModule.lastIndexOf('/')) !==
+            resolvedModule.substr(resolvedModule.lastIndexOf('/'))) {
+          moduleId = path.relative(this.file.fileName, resolvedModule)
+                         // path.relative starts from the parent directory (../self) but ES6
+                         // imports are relative to the current directory (./self) so we strip
+                         // one dot.
+                         .replace(/^\.\./, '.')
+                         // moduleResolution=node support: if importing from under
+                         // node_modules, omit the leading part of the path.
```

Comment:
We don't even run this code in google3 - the `convertIndexImportShorthand` option is set to false.

This only matches strings where TypeScript actually resolved a module. If google3 TypeScript were to resolve a file under a node_modules directory, we'd be in trouble already.